File: orchestrator.py
import asyncio
import logging

import yaml

from cache import APICache
from compactor import Compactor
from index import SeedIndex
from session import RateLimitGate, create_session
from worker import process_person


class Orchestrator:
    def __init__(self, config_path="config.yaml"):
        with open(config_path) as f:
            self.config = yaml.safe_load(f)

        rate_cfg = self.config.get("rate_limit", {})
        self.rate_gate = RateLimitGate(
            max_concurrent=rate_cfg.get("max_concurrent", 10),
            min_interval=rate_cfg.get("min_interval", 0.15),
            cooldown_base=rate_cfg.get("cooldown_base", 5.0),
            cooldown_max=rate_cfg.get("cooldown_max", 60.0),
        )

        self.cache = APICache(
            db_path=self.config["cache"]["db_path"],
            ttl_seconds=self.config["cache"]["ttl_seconds"],
            enabled=self.config["cache"]["enabled"],
            flush_interval=200,
        )
        self.compactor = Compactor(
            output_dir=self.config["output"]["dir"],
            found_suffix=self.config["output"]["found_suffix"],
            not_found_suffix=self.config["output"]["not_found_suffix"],
            years=self.config["filters"]["years_to_check"],
        )
        self.seed_index = SeedIndex()

    async def run(self, on_progress=None, on_log=None):
        if on_log:
            on_log("Iniciando procesamiento de datos")
        else:
            logging.info("Iniciando procesamiento de datos")

        max_workers = self.config["processing"]["max_workers"]
        rl_cfg = self.config.get("rate_limit", {})
        max_concurrent = rl_cfg.get("max_concurrent", max_workers)
        session, self.rate_gate = await create_session(limit=max_concurrent, rate_gate=self.rate_gate)

        try:
            files = self.seed_index.get_files()
            total_rows = sum(f["row_count"] for f in files)
            processed_overall = 0

            for file_info in files:
                found = []
                not_found = []
                msg = f"Procesando: {file_info['filename']} ({file_info['row_count']} filas)"
                if on_log: on_log(msg)
                else: logging.info(msg)

                batch_size = self.config["processing"]["batch_size"]
                processed_in_file = 0
                for start in range(0, file_info["row_count"], batch_size):
                    batch = self.seed_index.load_batch(
                        file_info["filepath"], start=start, size=batch_size
                    )
                    tasks = [
                        process_person(name, rfc, self.config, self.cache, session, self.rate_gate)
                        for name, rfc in batch
                    ]

                    results = await asyncio.gather(*tasks, return_exceptions=True)
                    for result in results:
                        if isinstance(result, Exception):
                            processed_overall += 1
                            continue
                        if result.get("Status") == "Found":
                            found.append(result)
                        else:
                            not_found.append(result)
                        processed_in_file += 1
                        processed_overall += 1

                    self.cache.flush()

                    if on_progress:
                        on_progress(processed_overall, total_rows)

                    msg = f"Lote completado: {processed_in_file}/{file_info['row_count']} registros"
                    if on_log: on_log(msg)
                    else: logging.info(msg)

                    inter_batch = self.config.get("rate_limit", {}).get("inter_batch_delay", 1.5)
                    await asyncio.sleep(inter_batch)

                summary = self.compactor.compact(found, not_found, file_info["basename"])
                msg = f"Completado {file_info['filename']}: {summary['found_count']} encontrados, {summary['not_found_count']} no encontrados"
                if on_log: on_log(msg)
                else: logging.info(msg)
        finally:
            await session.close()
            self.cache.close()

        msg = "Procesamiento completado."
        if on_log: on_log(msg)
        else: logging.info(msg)

File: src-tauri/src/orchestrator.rs
use crate::cache::ApiCache;
use crate::client::create_client;
use crate::compactor::Compactor;
use crate::models::AppConfig;
use crate::rate_limit::RateLimitGate;
use crate::seed_index::SeedIndex;
use crate::worker::process_person;
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use tokio::sync::mpsc;

pub struct Orchestrator {
    config: AppConfig,
    config_dir: PathBuf,
}

impl Orchestrator {
    pub fn new(config: AppConfig, config_dir: PathBuf) -> Self {
        Self { config, config_dir }
    }

    pub async fn run(
        &self,
        progress_tx: mpsc::UnboundedSender<crate::models::ProgressEvent>,
        log_tx: mpsc::UnboundedSender<crate::models::LogEvent>,
        stop_requested: Arc<AtomicBool>,
    ) -> Result<(), String> {
        let _ = log_tx.send(crate::models::LogEvent {
            message: "Iniciando procesamiento de datos".to_string(),
            level: "info".to_string(),
        });

        let cache_path = self.resolve_path(&self.config.cache.db_path);
        let output_dir = self.resolve_path(&self.config.output.dir);
        let seed_dir = self.config_dir.join("seed");

        let cache = ApiCache::new(
            &cache_path,
            self.config.cache.ttl_seconds,
            self.config.cache.enabled,
        )?;
        let client = create_client(&self.config)?;
        let rate_gate = RateLimitGate::new(
            self.config.rate_limit.max_concurrent,
            self.config.rate_limit.min_interval,
            self.config.rate_limit.cooldown_base,
            self.config.rate_limit.cooldown_max,
        );
        let mut seed_index = SeedIndex::new(&seed_dir)?;

        let files = seed_index.get_files().to_vec();
        let total_rows: usize = files.iter().map(|f| f.row_count).sum();
        let mut processed_overall: usize = 0;
        let mut found_overall: usize = 0;
        let mut not_found_overall: usize = 0;

        for file_info in &files {
            let mut found = Vec::new();
            let mut not_found = Vec::new();

            let _ = log_tx.send(crate::models::LogEvent {
                message: format!(
                    "Procesando: {} ({} filas)",
                    file_info.filename, file_info.row_count
                ),
                level: "info".to_string(),
            });

            let batch_size = self.config.processing.batch_size;
            let mut processed_in_file = 0;

            let mut stopped = false;

            for start in (0..file_info.row_count).step_by(batch_size) {
                let batch = seed_index.load_batch(
                    &PathBuf::from(&file_info.filepath),
                    start,
                    batch_size,
                )?;

                let mut tasks = Vec::new();
                for (name, rfc) in &batch {
                    let task = process_person(name, rfc, &self.config, &cache, &client, &rate_gate);
                    tasks.push(task);
                }

                let results = futures::future::join_all(tasks).await;
                for result in results {
                    match result.status.as_str() {
                        "Found" => found.push(result),
                        _ => not_found.push(result),
                    }
                    processed_in_file += 1;
                    processed_overall += 1;
                }

                cache.flush();

                let _ = progress_tx.send(crate::models::ProgressEvent {
                    processed: processed_overall,
                    total: total_rows,
                    found: found_overall + found.len(),
                    not_found: not_found_overall + not_found.len(),
                });

                let _ = log_tx.send(crate::models::LogEvent {
                    message: format!(
                        "Lote completado: {}/{} registros",
                        processed_in_file, file_info.row_count
                    ),
                    level: "info".to_string(),
                });

                if stop_requested.load(Ordering::Relaxed) {
                    let _ = log_tx.send(crate::models::LogEvent {
                        message: "Procesamiento detenido por el usuario.".to_string(),
                        level: "warn".to_string(),
                    });
                    stopped = true;
                    break;
                }

                let delay = self.config.rate_limit.inter_batch_delay;
                if delay > 0.0 {
                    tokio::time::sleep(std::time::Duration::from_secs_f64(delay)).await;
                }
            }

            let output_config = AppConfig {
                output: crate::models::OutputConfig {
                    dir: output_dir.to_string_lossy().to_string(),
                    found_suffix: self.config.output.found_suffix.clone(),
                    not_found_suffix: self.config.output.not_found_suffix.clone(),
                },
                filters: self.config.filters.clone(),
                ..self.config.clone()
            };
            let compactor = Compactor::new(&output_config);
            let summary = compactor.compact(&found, &not_found, &file_info.basename)?;

            found_overall += found.len();
            not_found_overall += not_found.len();

            let _ = log_tx.send(crate::models::LogEvent {
                message: format!(
                    "Completado {}: {} encontrados, {} no encontrados",
                    file_info.filename, summary.found_count, summary.not_found_count
                ),
                level: "info".to_string(),
            });

            if stopped {
                break;
            }
        }

        let _ = log_tx.send(crate::models::LogEvent {
            message: "Procesamiento completado.".to_string(),
            level: "info".to_string(),
        });

        Ok(())
    }

    fn resolve_path(&self, path: &str) -> PathBuf {
        let p = PathBuf::from(path);
        if p.is_absolute() {
            p
        } else {
            self.config_dir.join(path)
        }
    }
}
